### PR TITLE
[BACKLOG-5909] -- Added stateChanged event to the PromptPanel API

### PR DESCRIPTION
**Pull request by krivera-pentaho**

### CODE REVIEW

**Comment by krivera-pentaho** [comment]
@plagoa @diogofscmariano @annapopova1 @jvelasques @pamval please review

File: package-res/resources/web/prompting/PromptPanel.js [hunk 1]
```diff
@@ -446,6 +446,7 @@ define(['cdf/lib/Base', 'cdf/Logger', 'dojo/number', 'dojo/i18n', 'common-ui/uti
         onAfterRender: null,
         onBeforeUpdate: null,
         onAfterUpdate: null,
+        onStateChanged: null,
```
Comment (annapopova1) [inline package-res/resources/web/prompting/PromptPanel.js]: Need add property description (js doc) for 'onStateChanged' as for other class properties such as 'onBeforeRender'.

File: package-res/resources/web/prompting/api/EventAPI.js [hunk 2]
```diff
@@ -120,5 +120,21 @@ define([], function() {
         api.operation._getPromptPanel().ready = function(){};
       }
     };
+
+    /**
+     * Registers a listener for when the state of the prompt panel or parameter definition changes.
```
Comment (annapopova1) [inline package-res/resources/web/prompting/api/EventAPI.js]: Probably for users will be useful to describe all state flags which changes affect to stateChanged event. What do you think?

File: package-res/resources/web/test/prompting/PromptPanelSpec.js [hunk 3]
```diff
@@ -1363,6 +1363,24 @@ define([ 'dojo/number', 'dojo/i18n', 'common-ui/prompting/PromptPanel',
...
+
+          expect(panel.paramDefn).toBe(paramDefn);
+          expect(panel.onStateChanged).toHaveBeenCalledWith("promptNeeded", undefined, paramDefn.promptNeeded);
+          expect(panel.onStateChanged).toHaveBeenCalledWith("paginate", undefined, paramDefn.paginate);
+          expect(panel.onStateChanged).toHaveBeenCalledWith("totalPages", undefined, paramDefn.totalPages);
+          expect(panel.onStateChanged).toHaveBeenCalledWith("showParameterUI", undefined,paramDefn.showParameterUI());
+          expect(panel.onStateChanged).toHaveBeenCalledWith("allowAutoSubmit", undefined, paramDefn.allowAutoSubmit());
+          expect(panel.onStateChanged).toHaveBeenCalledWith("page", undefined, paramDefn.page);
+        });
+      });
```
Comment (annapopova1) [inline package-res/resources/web/test/prompting/PromptPanelSpec.js]: In addition better add one more test to check that onStateChanged callback is not called.

File: package-res/resources/web/prompting/api/EventAPI.js [hunk 4]
```diff
@@ -120,5 +120,21 @@ define([], function() {
...
+     * @method
+     * @param {Function} callback   The function to be executed when the event is triggered.
+     *                              Pass null if you wish to unbind this event.
+     * @example
+     *  api.event.stateChanged(function(name, oldValue, newValue) {
+     *    // Execute event based code
+     *  });
+     */
+    this.stateChanged = function(callback) {
+      api.operation._getPromptPanel().onStateChanged = callback;
```
Comment (annapopova1) [inline package-res/resources/web/prompting/api/EventAPI.js]: I would also validate that the callback is really function before assign it to onStateChanged

**Comment by annapopova1** [comment]
@krivera-pentaho What do you think about implemented PromptPanel#setState? Probably I should change it little bit to use method PromptPanel#setParamDefn that correctly call stateChanged callback.

**Comment by annapopova1** [comment]
One more comment about "autoSubmit" and "allowAutoSubmit" flags.
I read description BACKLOG-5909 and BACKLOG-6092 more attentively. Thomas separated those flags. I think the "allowAutoSubmit" flag matches to ParameterDefinition#allowAutoSubmit. And the "autoSubmit" flag matches to "autoSubmit" property from PromptPanel. And I think we need call callback for "autoSubmit" flag also. And accordingly I will change api.operation.state.
What do you think about it?

File: package-res/resources/web/prompting/api/EventAPI.js (hunk 2 above)
Comment (krivera-pentaho) [inline package-res/resources/web/prompting/api/EventAPI.js]: Definitely, I will add them

File: package-res/resources/web/prompting/api/EventAPI.js (hunk 4 above)
Comment (krivera-pentaho) [inline package-res/resources/web/prompting/api/EventAPI.js]: Not that i don't agree, but we are not doing it in any of the other calls, so we should do it for all.

**Comment by krivera-pentaho** [comment]
@annapopova1 if you use PromptPanel#setParamDefn then the onStateChanged would be called anyway. I don't think we need to explicitly call it in the PromptPanel#setState function, but it wouldn't hurt either.

Also, autoSubmit on the PromptPanel simply uses the "allowAutoSubmit" function value from the parameter definition. Is there somewhere else it is being set where I can trigger an onStateChanged call? It seems the only place is in the PromptPanel#setParamDefn call.

**Comment by krivera-pentaho** [comment]
@annapopova1 @plagoa @jvelasques @diogofscmariano @pamval please review

**Comment by annapopova1** [comment]
@krivera-pentaho allowAutoSubmit is initialized from server response and is set to paramDefn. PromptPanel#autoSubmit firstly is initialized from paramDefn but after that can be modified by submit component.

**Comment by krivera-pentaho** [comment]
@annapopova1 thank you for the clarification

**Comment by krivera-pentaho** [comment]
@annapopova1 @diogofscmariano @pamval @plagoa @jvelasques please review

**Comment by annapopova1** [comment]
1. What about "parametersChanged" property? According to BACKLOG-6092 api.operation.state allows modify this property. But "onStateChanged" works for this property only into PromptPanel#parameterChanged. So if we use api.operation.state to modify "parametersChanged" property, we do not trigger "onStateChanged" for it. Is it correct?

2. I do not sure about PromptPanel#setParamDefn because it each time overwrite PromptPanel#autoSubmit property by property from paramDefn. Before using setParamDefn method, PromptPanel#autoSubmit property is set single time in constructor. What do you think about it?

**Comment by krivera-pentaho** [comment]
@annapopova1 
1. I agree with this. I merged your PR so I will change that line to use setAutoSubmit
2. We definitely want to trigger onStateChanged when we use the state api call.
3. Do we want to change the prompt panel's property for being able to auto submit if we change the parameter definition, or should it only be changed the first time the parameter definition is set on the prompt panel?

**Comment by diogofscmariano** [comment]
@krivera-pentaho 
3. I think the property should not be changed in the first time :)
Apart from this agree

File: package-res/resources/web/prompting/PromptPanel.js [hunk 5]
```diff
@@ -531,9 +533,47 @@ define(['cdf/lib/Base', 'cdf/Logger', 'dojo/number', 'dojo/i18n', 'common-ui/uti
...
+            fireStateChanged("page", prevParamDefn, this.paramDefn, function(paramDefn) { return paramDefn.page; });
+          }
+        },
+
+        /**
+         * Sets the autoSubmit property on the PromptPanel
+         *
+         * @param autoSubmit {Boolean} The autoSubmit boolean
+         */
+        setAutoSubmit: function(autoSubmit) {
```
Comment (afrjorge) [inline package-res/resources/web/prompting/PromptPanel.js]: missing unit test

**Pull request merged**